### PR TITLE
Active flag

### PR DESCRIPTION
**Pull request by enriquefynn**
Steps towards validator's removal #347.
Replaces the validator's weight with the boolean flag `active`.
When the `active` flag is `false`, the validator cannot receive stake deposits. The maintenance bot should also ignore inactive validators and not try to stake more Sol to them.

### CODE REVIEW

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:34]: This needs to iterate all validators, not just the active ones. Right after if we flip the switch, the validator will still have active stake that is part of the Solido balance, it doesn’t disappear.

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:46]: The error message here might be misleading now, because the set of validators can be non-empty and we still return this error. If the error is not used anywhere else, we can rename the error.

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:81]: This line should be deleted.

File: program/src/state.rs
Comment (ruuda) [inline program/src/state.rs:633]: With booleans, it is often simpler to avoid a double negation, then you don’t have to work all of that out in your head. In this case, we can change the field to `active: bool`, and have it default to true.

Some evidence for this, is that almost all places in the code where this field is referenced, now write `!validator.inactive`, and then you have to mentally make the step “not inactive ⇒ active”. It would be a bit simpler to just write `validator.active` in those places.

File: program/src/state.rs
Comment (ruuda) [inline program/src/state.rs:640]: Let’s also initialize the field explicitly (maybe let’s just initialize all fields explicitly and remove the `Default`), I find it a bit scary to rely on an implicit default for such an important field.

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:238]: If we set this to a non-zero value, that would have caught the bug above where inactive validators are not considered for the current balance.

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:257]: This comment is not correct.

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:240]: We should keep the non-integer multiple test, it was testing the distribution of the additional lamports at the end. In the new test above, with 400 lamports and 2 active validators, the total is a multiple of the number of validators, so there is no remainder to divide. We should also test the case where the total is not a multiple, and the target of some validators is therefore larger than that of others.

It should be possible to see this in the code coverage results too. You can run `tests/coverage.py` locally, and that will produce an html report where you can see which lines of the distribution logic are covered.

**Review by ruuda — COMMENTED** [review]
The approach looks good! I have a few comments about the details.

File: program/src/state.rs
Comment (enriquefynn) [reply program/src/state.rs:633]: I originally did it like that, but then I'd lose the default `entry` constructor from `AccountMap`. But we could also implement the custom default for a `Validator`.

File: program/src/balance.rs
Comment (enriquefynn) [reply program/src/balance.rs:240]: Oops... will do that, thanks!

Comment (enriquefynn) [reply program/src/balance.rs:240]: Wouldn't this test be equal to the `get_target_balance_works_for_non_integer_multiple`?

File: program/src/state.rs
Comment (ruuda) [reply program/src/state.rs:633]: :+1: Let’s implement `Default` manually for `Validator` then.

File: program/src/balance.rs
Comment (ruuda) [reply program/src/balance.rs:240]: There is a small difference: when one of the validators is inactive, distributing the remainder should skip over those. To test it properly, we need to make sure then that the inactive validator(s) are early in the list, such that there is a validator behind an inactive one, that gets one additional lamport.

**Comment by ruuda** [comment]
Let’s not merge this to `main` yet, with the testnet dry-run going on. At the moment, `main` is still compatible with the program that is deployed to devnet, but this would break that. We can make a new branch `next` or something like that for now, and continue development of v0.4.0-incompatible features on that branch. Then after the testnet run is complete, we can merge it into `main`.

**Comment by enriquefynn** [comment]
> Let’s not merge this to main yet, with the testnet dry-run going on. At the moment, main is still compatible with the program that is deployed to devnet, but this would break that. We can make a new branch next or something like that for now, and continue development of v0.4.0-incompatible features on that branch. Then after the testnet run is complete, we can merge it into main.

Great, I modified the base branch, let's do it like that.

File: program/src/balance.rs
Comment (ruuda) [inline program/src/balance.rs:250]: s/validator/validators/

File: program/src/state.rs
Comment (ruuda) [inline program/src/state.rs:630]: s/true/false/

**Review by ruuda — APPROVED** [review]
Found two typos, aside from those, LGTM! :shipit:

**Pull request merged**
